### DISCUSSION TITLE
Instrumentation: code coverage

### DISCUSSION THREAD

**Discussion by galdolber** [borkdude/sci#807] [Q&A]
I'm evaluating different solutions to implement code coverage for code I'm running with sci.

Is there any way to use the interpreter to achieve this kind of instrumentation?
Some way to hook into the evaluation process.

**Sponsor**

**Comment by borkdude** [comment]
@galdolber Thank you so much for sponsoring, I really appreciate that.

Currently there aren't any hooks that would help with this. Perhaps this could be done with macros. I'm not deeply familiar with how other solutions do this in Clojure. What are your alternatives so far?

**Reply by galdolber** [reply]
So far this is what I found:
https://github.com/cloverage/cloverage will rewrite the code and wrap with the instrumentation code
https://github.com/technomancy/radagast more simplistic implementation, looks like it wraps vars and uses metadata to tracks access to them

Didn't try to run or migrate any of them to sci yet.
Cloverage's approach is good, but given that sci is an interpreter I have the feeling something much better can be done.
I've been looking at sci's code, If I were to try to add these kind of hooks, any clues where I should start?

**Reply by borkdude** [reply]
SCI vars are similar to Clojure vars: they can be instrumented like instrumentation in `clojure.spec.test.alpha` works by replacing the root binding with a function that "wraps" using `(sci/)alter-var-root`. This would be the easiest implementation that requires no changes to SCI.

When adding things to SCI for instrumentation, probably the `analyzer.cljc` file is the place to look and add instrumentation stuff to the nodes that are generated and executed at runtime.

**Reply by borkdude** [reply]
I was able to hack cloverage + bb locally to make _something_ work:

<img width="1987" alt="Screenshot 2022-12-25 at 13 32 40" src="https://user-images.githubusercontent.com/284934/209468095-afbce3d6-3533-426f-a03f-1caa5c03b3e8.png">
<img width="750" alt="Screenshot 2022-12-25 at 13 32 31" src="https://user-images.githubusercontent.com/284934/209468097-d9d3237f-9c08-476d-95a6-9ef93bf6214c.png">
<img width="1222" alt="Screenshot 2022-12-25 at 13 32 25" src="https://user-images.githubusercontent.com/284934/209468098-e6316287-03f1-4093-8950-c269e9c28606.png">

It requires a branch of bb called `tools-analyzer` and the branch `cloverage-bb` that I pushed here:

https://github.com/borkdude/cloverage/tree/cloverage-bb

Don't know if your use case was babashka or more SCI in general

**Comment by borkdude** [comment]
A follow up:

In the latest babashka / SCI (master) `loop*`, `fn*` are now special forms and `loop`, `fn`, `defn`, `defmacro`, and `and` and `or` are implemented as macros, which is a step in the right direction for this use case.

The [rcf](https://github.com/hyperfiddle/rcf) library has a kind of tools.analyzer-like thing going on in their code which is compatible with the latest babashka. It could be an interesting way to enable code coverage tools as well.

**Comment by borkdude** [comment]
With a few hacks I was able to make tools.analyzer work in bb:

https://github.com/babashka/babashka/issues/1441

Perhaps that will also help this use case.
